src/Test/Hspec/Expectations.hs

```diff
@@ -88,6 +89,23 @@ with_loc(shouldBe, (Show a, Eq a) => a -> a -> Expectation)
 actual `shouldBe` expected = expectTrue ("expected: " ++ show expected ++ "\n but got: " ++ show actual) (actual == expected)
 
 -- |
+-- @actual \`shouldBeNear\` expected@ sets the expectation that @actual@ be a
+-- floating point value near @expected@. If either value is zero we check that
+-- the absolute difference is less than epsilon (1e-15) otherwise we check if
+-- the relative difference is less than epsilon.
+with_loc(shouldBeNear, (Show a, Ord a, Floating a) => a -> a -> Expectation)
+actual `shouldBeNear` expected
+    -- Short circuit if they are actually equal.
+    | actual == expected = reportFail (actual == expected)
+    | actual == 0 || expected == 0 = reportFail (absoluteDifference < epsilon)
+    | otherwise =  reportFail (relativeDifference < epsilon)
```

Comment:
> I haven't tried, but what happens if x is
>
>    1. `0.9e-300`
>    2. `0`
>
> would (1) fail and (2) pass?

That is what currently happens, and I agree, at first sight, doesn't make much sense, but this is the usual recommended way to compare floating point numbers.

However, now that I think about it,
```
1e-300 `shouldBeNear` 0.0
```
should pass (because `actual` is close to `expected`), but I think maybe
```
0.0 `shouldBeNear` 1e-300
```
should fail.

Let me think about this some more.